Option to configure when the error sound may be played.

CyrilleB79 opened the pull request:
### Link to issue number:

#12672

### Summary of the issue:
The error sound that is played in test versions may also be useful to testers on stable and RC releases.
The motivation to have this feature included in NVDA's core rather than in an add-on is that it may be useful to test NVDA stable versions with add-ons disabled.

### Description of how this pull request fixes the issue:

Added a 'Play a sound for logged errors' option in the advanced settings panel. It is a combo-box with two choices:
- 'Only in test versions': NVDA plays error sounds as today, i.e. only in NVDA test versions (alpha, beta and from source) but not in RC or stable.
- Yes: NVDA plays the error sound whatever its version (test or stable)
According to the discussion in #12672, I do not provide an option to disable error sounds in tests versions, since test versions are aimed to capture potential issues. Thus error notifications should not be disabled for this usage.

If an error occurs before the config is initialized, the default behaviour is applied, i.e. the error sound is played only in test versions.

Note for reviewers:
* I have had to import config in the `handle` method rather than at the top of the module since NVDA could not start in the latter case. I do not know why however, since I do not know anymore how to debug this since the virtual environment has been set up.
* Please native English speakers, check my writings (GUI and user guide)

Additional modification:
While working on this part of the code, I have made the critical error sound played asynchronously upon @leonardder's suggestion. It should allow to spare 2-3 seconds (the time of the error sound) in the code execution time.
NV Access, if you prefer not to include this modification for whatever reason, I or you may revert this commit.

### Testing strategy:

Manual testing:
Triggered an error in the 4 following situations and checked that the error sound is played for the 3 first ones and not the last one:
* NVDA test version (from source) and option set to 'Yes'
* NVDA test version (from source) and option set to 'Only in NVDA test versions'
* NVDA stable version (simulated from source) and option set to 'Yes'
* NVDA stable version (simulated from source) and option set to 'Only in NVDA test versions'

The stable version is simulated running NVDA from source with this branch and entering the 2 following lines in the Python console:
```
import buildVersion
buildVersion.isTestVersion = False
```

The error is triggered by trying to enable the screen curtain while Windows Magnifier is enabled with color inversion activated. But many other ways could have been used.

I will also check context help on the snapshot generated by appVeyor.

### Known issues with pull request:
If an error occurs before the config is initialized, the default behaviour is applied, i.e. the error sound is played only in test versions whatever the value of this option is in the config.

### Change log entries:
New features (or Changes? Or For developers?)
It is now possible to configure the error sound to be played in any version of NVDA, not only in test versions. (#12672)

### Code Review Checklist:

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual testing.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

CyrilleB79 committed:
Option to configure when the error sound may be played.

CyrilleB79 committed:
Lazy import of config.

lukaszgo1 commented on source/logHandler.py:
```diff
@@ -266,8 +266,13 @@ def emit(self, record):
-		# Only play the error sound if this is a test version.
-		shouldPlayErrorSound =  buildVersion.isTestVersion
+		import config
+		# Only play the error sound if this is a test version or if the config states it explicitely.
```
explicitely -> explicitly
```suggestion
		# Only play the error sound if this is a test version or if the config states it explicitly.
```

lukaszgo1 commented on source/logHandler.py:
```diff
@@ -266,8 +266,13 @@ def emit(self, record):
+		shouldPlayErrorSound = (
+			buildVersion.isTestVersion
+			# Play error sound: 1 = Yes
+			or config.conf["featureFlag"]["playErrorSound"] == 1
```
You need to protect against `config.conf` being `None` here. This can happen if anything is logged at the error level between `logHandler` being initialized and the call to `config.initialize` in `core.main`.

CyrilleB79 replied:
Thank you. Fixed in 9a1cb16.
I'll update the initial description regarding the inability to have the error sound played in the stable release before the log is initialized.

CyrilleB79 committed:
Linting

CyrilleB79 committed:
Fix error when log is not yet initialized.

CyrilleB79 committed:
Fixed typo

Co-authored-by: Łukasz Golonka <lukasz.golonka@mailbox.org>

CyrilleB79 commented:
I have checked context help on the appVeyor artifact. And I have addressed @lukaszgo1's comments (thanks Lukasz for the review).
Thus I have passed this PR to ready state.

lukaszgo1 commented:
Looking at this again would it be possible to move the code which is responsible for determining if the error sound should play into a separate method/function? It would allow add-on developers / power users to monkey patch just a single function rather than the entire `emit` and would allow to indirectly add 'never' option for developers.

CyrilleB79 committed:
Make shouldPlayErrorSound a function.

CyrilleB79 commented:
> Looking at this again would it be possible to move the code which is responsible for determining if the error sound should play into a separate method/function? It would allow add-on developers / power users to monkey patch just a single function rather than the entire `emit` and would allow to indirectly add 'never' option for developers.

Done.

leonardder commented on source/logHandler.py:
```diff
@@ -266,14 +281,12 @@ def emit(self, record):
-		shouldPlayErrorSound =  buildVersion.isTestVersion
 		if record.levelno>=logging.CRITICAL:
 			try:
 				winsound.PlaySound("SystemHand",winsound.SND_ALIAS)
```
May be you could fix this to play asynchronously while at it.

```suggestion
				winsound.PlaySound("SystemHand", winsound.SND_ALIAS | winsound.SND_ASYNC)
```

CyrilleB79 committed:
Play asynchronously critical error sound

While modifying this part of code, let's make critical error sound playing asynchronous - allows to spare a few seconds in the code execution. Thanks @leonardder for suggestion.

Co-authored-by: Leonard de Ruijter <leonardder@users.noreply.github.com>

CyrilleB79 commented:
This PR is ready again.

@feerrenrut you may want to review this one since you gave indications in https://github.com/nvaccess/nvda/issues/12672#issuecomment-885357649.

feerrenrut reviewed (approved):
The only thing that concerns me is the async sound playing. It is probably fine, and I'm happy for this to get some testing on alpha. But if there is an issue, we'll need to revert that aspect.

I won't merge this immediately, to give others a chance to review.

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into beepOnError

feerrenrut committed:
Update changes file for PR #12689
